c/tskit/trees.c

```diff
@@ -7644,3 +7644,417 @@ tsk_treeseq_divergence_matrix(const tsk_treeseq_t *self, tsk_size_t num_samples,
+            while ((p != TSK_NULL) && (degree[p] == 0) && (p != p_in) && not_sample[p]) {
+                p = out_parent[p];
+            }
+            if (p == p_in) {
+                // we can extend!
+                // But, we might have passed the interval that a
+                // postponed edge in covers, in which case
+                // we should skip postponing the edge in
+                if (far_side[e_in] != there) {
+                    ex_in->extended = true;
+                }
+                near_side[e_in] = there;
+                while (c != p) {
+                    for (ex_out = edges_out_head; ex_out != NULL;
+                         ex_out = ex_out->next) {
+                        e_out = ex_out->edge;
+                        if (edges->child[e_out] == c) {
+                            break;
+                        }
+                    }
+                    tsk_bug_assert(edges->child[e_out] == c);
+                    ex_out->extended = true;
+                    far_side[e_out] = there;
+                    // amend degree: the intermediate
+                    // nodes have 2 edges instead of 0
+                    tsk_bug_assert(degree[c] == 0 || c == edges->child[e_in]);
+                    if (degree[c] == 0) {
+                        degree[c] = 2;
+                    }
+                    c = out_parent[c];
+                }
+            }
+        }
+        if (forwards) {
+            valid = tsk_tree_position_next(&tree_pos);
+        } else {
+            valid = tsk_tree_position_prev(&tree_pos);
+        }
+    }
+
+    for (e = 0; e < (tsk_id_t) num_edges; e++) {
+        keep[e] = edges->left[e] < edges->right[e];
+    }
+    ret = tsk_edge_table_keep_rows(edges, keep, 0, NULL);
+out:
+    tsk_blkalloc_free(&edge_list_heap);
+    tsk_tree_position_free(&tree_pos);
+    tsk_safe_free(degree);
+    tsk_safe_free(out_parent);
+    tsk_safe_free(keep);
+    tsk_safe_free(not_sample);
+    return ret;
+}
+
+static int
+tsk_treeseq_slide_mutation_nodes_up(
+    const tsk_treeseq_t *self, tsk_mutation_table_t *mutations)
+{
+    int ret = 0;
+    bool valid;
+    double t, right;
+    tsk_id_t c, p, tj, e, next_mut;
+    tsk_tree_position_t tree_pos;
+    const tsk_table_collection_t *tables = self->tables;
+    const tsk_size_t num_nodes = tables->nodes.num_rows;
+    const tsk_edge_table_t *edges = &tables->edges;
+    tsk_id_t *parent = tsk_malloc(num_nodes * sizeof(*parent));
+    double *sites_position = tables->sites.position;
+    double *nodes_time = tables->nodes.time;
+
+    tsk_memset(&tree_pos, 0, sizeof(tree_pos));
+
+    if (parent == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    tsk_memset(parent, 0xff, num_nodes * sizeof(*parent));
+
+    ret = tsk_tree_position_init(&tree_pos, self, 0);
+    if (ret != 0) {
+        goto out;
+    }
+
+    valid = tsk_tree_position_next(&tree_pos);
+    next_mut = 0;
+
+    while (valid) {
```

Comment:
Some comment as Python code here - why not just use the standard tree API? Any perf difference will be neligible.